feat: local c2st metric

JuliaLinhart opened the pull request:
## What does this implement/fix? Explain your changes

L-C2ST(-NF) diagnostic: class, tutorial and tests.

## Does this close any currently open issues?

Fixes #1005 

## Any relevant code examples, logs, error output, etc?

...

## Any other comments?

...

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask. We're here to
help! This is simply a reminder of what we are going to look for before merging
your code.

- [`x` ] I have read and understood the [contribution
  guidelines](https://github.com/sbi-dev/sbi/blob/main/CONTRIBUTING.md)
- [`x` ] I agree with re-licensing my contribution from AGPLv3 to Apache-2.0.
- [ `x`] I have commented my code, particularly in hard-to-understand areas
- [`x` ] I have added tests that prove my fix is effective or that my feature works
- [ `x`] I have reported how long the new tests run and potentially marked them
  with `pytest.mark.slow`.
- [ `x`] New and existing unit tests pass locally with my changes
- [`x` ] I performed linting and formatting as described in the [contribution
  guidelines](https://github.com/sbi-dev/sbi/blob/main/CONTRIBUTING.md)
- [ ] I rebased on `main` (or there are no conflicts with `main`)

agramfort commented on sbi/analysis/plot.py:
```diff
@@ -1403,3 +1408,166 @@ def _plot_hist_region_expected_under_uniformity(
+            pp_vals_null[t] = pd.Series(PP_vals(probas_null[t], alphas))
+
+    low_null = pd.DataFrame(pp_vals_null).quantile(conf_alpha / 2, axis=1)
+    up_null = pd.DataFrame(pp_vals_null).quantile(1 - conf_alpha / 2, axis=1)
```
maybe you don't need pandas here? you can use https://numpy.org/doc/stable/reference/generated/numpy.quantile.html
I don't know if pandas is already a dependency of sbi but it's always a good idea to limit dependencies when they can be avoided (think import time, CI issues on updates)

JuliaLinhart replied:
I'm fairly certain that pandas it is a dependency, but I agree. The small fixes you suggested are totally doable. However I would also have to replace the `.groupy` method from pandas, which will probably require to implement a `numpy_groupy` function...

agramfort replied:
you can use https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binned_statistic.html#

but now if Pandas is already used everywhere I think it's fair to ignore my point...

agramfort commented on sbi/analysis/plot.py:
```diff
@@ -1403,3 +1408,166 @@ def _plot_hist_region_expected_under_uniformity(
+# ==== Diagnostics for L-C2ST ====
+
+
+def pp_plot_lc2st(
```
I would not put lc2st in the function name. This function is a generic pp plot function.

JuliaLinhart replied:
I agree. I will create a new function for general PP-plots and call it inside the `lc2st_pp_plot` function. Some things like the theoretical null values (here the dirac at 0.5) change depending on the considered test.

agramfort commented on sbi/analysis/plot.py:
```diff
@@ -1403,3 +1408,166 @@ def _plot_hist_region_expected_under_uniformity(
+    )
+
+    for p, l_, c in zip(probas, labels, colors):
+        pp_vals = pd.Series(PP_vals(p, alphas))
```
idem. Series is not necessary here

agramfort commented on sbi/analysis/plot.py:
```diff
@@ -1403,3 +1408,166 @@ def _plot_hist_region_expected_under_uniformity(
+    n_bins=20,
+    vmin=0,
+    vmax=1,
+    cmap=cm.get_cmap("Spectral_r"),
```
```suggestion
    cmap="Spectral_r",
```
should be enough. You can call cm.get_cmap inside the function.

agramfort commented on sbi/analysis/plot.py:
```diff
@@ -1403,3 +1408,166 @@ def _plot_hist_region_expected_under_uniformity(
+        _, bins, patches = ax.hist(df_probas.z, n_bins, density=True, color="green")
+        df_probas["bins"] = np.select(
+            [df_probas.z <= i for i in bins[1:]], list(range(n_bins))
+        )
```
I think you should be able to do this with np.searchsorted

agramfort commented on sbi/analysis/plot.py:
```diff
@@ -1403,3 +1408,166 @@ def _plot_hist_region_expected_under_uniformity(
+            [df_probas.z <= i for i in bins[1:]], list(range(n_bins))
+        )
+        # get mean predicted proba for each bin
+        weights = df_probas.groupby(["bins"]).mean().probas
```
you can also do this with https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binned_statistic.html

I have nothing agains pandas but I feel that numeric package should it as much as possible. Doing pure numpy / scipy operations will generally be faster.

agramfort commented on sbi/analysis/test_utils.py:
```diff
@@ -0,0 +1,50 @@
+import pandas as pd
+
+
+def PP_vals(RV_samples, alphas):
```
you should avoid capitals in function name and variable names (cf pep8 conventions)
```suggestion
def pp_vals(rv_samples, alphas):
```

agramfort commented on sbi/simulators/gaussian_mixture.py:
```diff
@@ -0,0 +1,121 @@
+import logging
```
I would consider splitting this PR with a first PR for this GMM simulator.

agramfort replied:
it would also force you to add a test for the simulator :)

JuliaLinhart replied:
Good Idea, but I won't have the time. @michaeldeistler , @janfb what do you think?

janfb replied:
Do we really need the Gaussian mixture simulator for the test? If so, I would add it analogously to the `linear_gaussian` simulator, i.e., just add the simulator as plain function and then add another function that can samples the reference posterior. The prior will be defined in the test then.

Also, the class looks very similar to the sbibm implementation. If we use that here we should add a note.

JuliaLinhart replied:
I used the `gaussian_mixture` in the tutorial as an example and it was therefore easier for me to use it in the tests again. I also think it's a useful example more generally speaking for future tutorials. So I will keep it and make the changes you requested @janfb. 

If you want I can add a test with the `gaussian_linear` simulator too. Let me know!

agramfort commented on tests/lc2st_test.py:
```diff
@@ -0,0 +1,220 @@
+    num_simulations = 100
+    max_num_epochs = 1
+
+    n_cal = 100
+    n_eval = 100
+    num_trials_null = 2
```
the mix of notations n_ vs num_ seems inconsistent

agramfort commented on tests/lc2st_test.py:
```diff
@@ -0,0 +1,220 @@
+
+@pytest.mark.slow
+@pytest.mark.parametrize("method", (LC2ST, LC2ST_NF))
+def test_lc2st_TPR(method):
```
```suggestion
def test_lc2st_TPR(method):
```
```suggestion
def test_lc2st_tpr(method):
```

agramfort commented on tests/lc2st_test.py:
```diff
@@ -0,0 +1,220 @@
+
+
+@pytest.mark.parametrize("method", (LC2ST, LC2ST_NF))
+def test_lc2st_TNR(method):
```
```suggestion
def test_lc2st_tnr(method):
```

agramfort commented on tests/lc2st_test.py:
```diff
@@ -0,0 +1,220 @@
+
+    # task
+    gmm = GaussianMixture()
+    prior = gmm.prior_dist
```
here you overwrite the prior parameter of the test

JuliaLinhart commented:
Thanks for this review! I'll do the changes, review the doc and fix typing issues.

JuliaLinhart commented:
Almost all the suggestions by @agramfort have been addressed. Except:
- `pandas` is still used for the `.groupy()` method in the `marginal_plot_with_proba_intensity` function from `sbi.analysis.plot.py`.
- the added simulator has no corresponding `pytest` script.

Future fixes include:
- function that regroups the `marginal_plot_with_proba_intensity` into a single pairplot. 
- generic `HypothesisTest` class to centralize `sbc`, `lc2st` and other diagnostics relying on hypothesis testing.

janfb commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,730 @@
+        else:
+            return scores
+
+    def train_data(self) -> None:
```
I am confused by this method and its name `train_data`. 

Is it something `sklearn` related? Why not call `_train` directly? or just call it `train`? 

Or, if it needed as a wrapper, call it `train_on_observed_data`?

JuliaLinhart replied:
Again this relates to the comment above about training the classifiers on observed data or under the null hypothesis. I changed the name of the methods to a more explicit ones (`train_on_observed_data` and `train_under_null_hypothesis`)

janfb commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,730 @@
+        self.trained_clfs = trained_clfs
+
+    def scores_data(
+        self, P_eval: Tensor, x_eval: Tensor, return_probas: bool = False
```
Some renaming suggestions to conform with our conventions: 
```suggestion
        self, theta_o: Tensor, x_o: Tensor, return_probs: bool = False
```

or use `posterior_samples` instead of `theta_o`.

janfb commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,730 @@
+        trained_clfs = self._train(self.P, self.Q, self.x_P, self.x_Q)
+        self.trained_clfs = trained_clfs
+
+    def scores_data(
```
Again, I am confused by the name here :) maybe just call it `scores`? or `get_scores` to use a verb?

JuliaLinhart replied:
Same as before. There is a difference to be made between computing the scores on observed data or under the null hypothesis. I changed the name of the method to a more explicit one (`scores_on_observed_data`). Computing the scores under the null hypothesis is integrated in the method `statistics_null` that i changed to `statistics_under_null_hypothesis` to be more explicit.

JuliaLinhart replied:
I made the decision of not having a `scores_under_null_hypothesis` method because I don't thing this is needed.

janfb commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,730 @@
+        """Compute the L-C2ST scores for the observed data.
+
+        Args:
+            P_eval: Samples from P conditioned on the observation `x_eval`,
```
Let's be more explicit here: 
```suggestion
            theta_o: Samples from the posterior conditioned on the observation `x_o`,
```

janfb commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,730 @@
+            return_probas=return_probas,
+        )
+
+    def statistic_data(
```
`get_statistics`? :)

janfb commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,730 @@
+
+    def statistic_data(
+        self,
+        P_eval: Tensor,
```
see comments above regarding naming.

janfb commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,730 @@
+            x_eval: Observation, of shape (, dim_x).
+
+        Returns:
+            p-value for L-C2ST at `x_eval`.
```
below we take a `mean()`. I think it would be good to mention here, how the stats are averaged to get the `p_value`.

janfb commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,730 @@
+        )
+        return (stat_data < stats_null).mean()
+
+    def reject(
```
I suggest to make it very explicit: 
```suggestion
    def p_value_smaller_alpha(
```

then you can things like 
```python
if p_value_smaller_alpha:
    ...
```

JuliaLinhart replied:
As you which. This is just classical hypothesis test vocabulary. I changed it for now, we can always revert.

janfb commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,730 @@
+        # Draw samples from the base distribution for evaluation
+        self.P_eval = flow_base_dist.sample(torch.Size([n_eval]))
+
+    def _scores(
```
maybe I am missing something, but to me it seems all the class methods below are just calling `super.method(...)`. 
Unless there is something flow-related that needs to be changed, you could just delete them here, no?

JuliaLinhart replied:
They use the samples from the base distribution for `P_eval` (or now `theta_o`). So basically you are right, the methods call the parent method but with specified `self.P`. This is just to avoid misusing the NF-method and passing wrong `P_eval`.

JuliaLinhart replied:
No, there is no other scenario, because the base distribution from which we get the new `theta_o` is independent of `x_o`. But I can make it the user's responsibility. You tell me!

janfb replied:
I am not sure I understand. In the parent class, `theta_o` are samples from the posterior given `x_o`, but here this changes so that `theta_o` are samples from the underlying base distribution? I think this would need a bit more comments and mentioning in the docstrings. At the moment, the class docstring says that samples from the prior and the posterior (i.e., `theta_o`) are inverse transformed and evaluated under the `base`, but then `theta_o` is initialized from the `base` in `__init__`?

janfb replied:
Can you please help here @JuliaLinhart? I am missing something, or was this resolved?

JuliaLinhart replied:
Okay so basically a sample from the flow is obtained by pushing a sample from the base distribution through the forward flow transformation: $\theta = T_\phi(z; x_o) \sim q_\phi(\theta \mid x_o)$. So applying the inverse transformation to this sample, recovers the sample from the base distribution. In the end, the `LC2ST_NF` method compares samples from the base distribution $z_{c=0} \sim \mathcal{N}(0,1)$ (class $c=0$) and samples from the true posterior $\theta_o \sim p(\theta \mid x_o)$ inversely transformed by the flow (class $c=1$):
$$z_{c=1} = T_\phi^{-1}(\theta_o; x_o) \sim p(T_\phi^{-1}(\theta; x_o) \mid x_o)$$

JuliaLinhart replied:
Does this clarify things? You can also look at the tutorial or the paper itself, where this `-NF` version is explained more in detail. 

Let me know if I should change the code and / or add a description.

janfb replied:
ok, make sense now. thanks for clarifying!

janfb commented on tests/lc2st_test.py:
```diff
@@ -0,0 +1,217 @@
+def test_running_lc2st(method, classifier, cv_folds):
+    """Tests running inference and then LC2ST-(NF) and then getting test quantities."""
+
+    n_train = 100
```
pleaser use similar names as in other tests, e.g., `num_simulations`, `num_{cal, eval, trials_null}`

janfb reviewed (commented):
Awesome, thanks a lot for adding this! 
Looks great already, I added a couple of comments and questions. I think this needs some renaming here and there to match our variable names conventions and PEP 8.

JuliaLinhart commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,730 @@
+        if "mlp" in classifier.lower():
+            ndim = thetas.shape[-1]
+            self.clf_class = MLPClassifier
+            self.clf_kwargs = {
```
I've added the `if None` statement. The `clf_kwargs` was already there.

JuliaLinhart commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,730 @@
+        if self.n_folds > 1:
+            trained_clfs = []
+            kf = KFold(n_splits=self.n_folds, shuffle=True, random_state=self.seed)
+            cv_splits = kf.split(self.P)
```
My bad, this should be performed on `P` not `self.P`. I fixed it.

The problem is that this function is also used for training the classifiers under the null hypothesis, so for different `P`. Whether to use `self.P` or the permuted `P_perm` is specified in the `train_data` and `train_null` functions.

I prefer to leave it like that, as the data must be either the `self.P,Q,x_P,x_Q` or the permuted versions. There is no other option.

JuliaLinhart commented:
I think I've addressed all the comments @janfb. Let me know if there's something else I can do :)

JuliaLinhart commented:
I think I have addressed all your comments and requests @janfb, except the one where I should get rid of the `groupby` method from `pandas`. I will try to fix this as soon as I can.

JuliaLinhart commented:
All done @janfb

JuliaLinhart commented on sbi/analysis/test_utils.py:
Oh of course, I've changed it to live in `sbi/utils/analysis_utils.py`.

psteinb commented:
I am happy to help review this PR. But given the activity that is already visible, I'd push this effort to a later stage. Feel free to ping me if my help is needed.

JuliaLinhart commented on tests/lc2st_test.py:
```diff
@@ -0,0 +1,224 @@
+
+
+@pytest.mark.parametrize("method", (LC2ST, LC2ST_NF))
+def test_lc2st_tnr(method):
```
I made the name more explicit: `true_positif/negative_rate`. I also changed the test slightly to actually test the "TPR/TNR", i.e. how many times we correctly accept or reject the test over 100 runs. It takes some time, but this will be a "slow" test.

JuliaLinhart commented:
Response to review from @janfb: the above commit fixes following requests
- rename tutoral to `18_...`, plots and results description 
- change content of `anamysis/test_utils.py` to `sbi/utils/analysis_utils.py`
- description of p-value computation in `lc2st.py`
- explicit name and description of the tests in `lc2st_tests.py` and adapt code to test the true  postitive and negative rates of the hypothesis test. Runtime is longer for 100 test runs (7min instead of less than a minute for 10 runs), but otherwise the "rate" is not a trusworthy empirical result in my opinion)

janfb commented on tests/lc2st_test.py:
```diff
@@ -0,0 +1,245 @@
+
+@pytest.mark.slow
+@pytest.mark.parametrize("method", (LC2ST, LC2ST_NF))
+def test_lc2st_true_positif_rate(method):
```
typo:
```suggestion
def test_lc2st_true_positiv_rate(method):
```

janfb reviewed (commented):
Thanks for the updates - looks good! 

One more thing, the `theta_o` generation in the `LC2ST_NF` needs clarification.

janfb commented:
> I am happy to help review this PR. But given the activity that is already visible, I'd push this effort to a later stage. Feel free to ping me if my help is needed.

That's great @psteinb ! Do you have capacity to review the tutorial? That'd be great 🙏

JuliaLinhart commented:
I think the `npe.sample` method (where `npe` is a `DensityEstimator` object) is different for the tests vs. on the branch I am working on. It seems to be the handling of the `context` variable, but I am not sure because I can't verify anything locally :/

michaeldeistler commented:
just rebase on main
```
git checkout main
git pull
git checkout 1005-implement-l-c2st-metric
git rebase main
```

JuliaLinhart commented:
Oh right.. Sorry! So I fixed it, but had to reshape a lot...

JuliaLinhart commented:
I did some experiments and I was wondering how you choose the parameters for the`MLPClassifier` from `sklearn` corresponding to the default `classifier=mlp` in `c2st` and the `LC2ST` class. Especially `early_stopping` and the regularization parmater `alpha` seem to be a limitation in some application examples... @psteinb you worked on that right?

For me the `MLPClassifier` from `sklearn` with default parameters with `alpha=0` and  `max_iter=25000` yields pretty stable results, but is prone to overfitting. I therefore suggest that if we stick with your `mlp`, to do ensembling with different seeds (over 5 models  by default). It yields more stable results with smaller confidence regions, but is slower. This is my last commit and I added tests. ]

Let me know what you think :)

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+   "source": [
+    "##### PP-Plots\n",
+    "\n",
+    "PP-plots allow to evaluate a general trend of over or under confidence, by comparing the predicted probabilities of belonging to the estimated posterior (class 0). If the red curve is not fully contained in the gray confidence region, this means that the test rejects the null hypothesis and that significant deviances from the true postrior are detected. Here two scenarios are possible:\n",
```
```suggestion
    "P-P plots allow to evaluate a general trend of over- or under-confidence, by comparing the predicted probabilities of belonging to the estimated posterior (class 0). If the red curve is not fully contained in the gray confidence region, this means that the test rejects the null hypothesis and that a significant discrepancy from the true posterior is detected. Here two scenarios are possible:\n",
```

psteinb replied:
I suggest to make the statement at the end more strong. discrepancy can do the job for me.

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+    "        ax=axes[i],\n",
+    "    )\n",
+    "    axes[i].set_title(f\"PP-plot for observation {i+1}\")\n",
+    "axes[i].legend()\n",
```
Note, the confidence region does not contain the true null. In the paper's figure 4, this is the case. It might cause some confusion that the dashed line in these plots is not contained in the grey CI area.

JuliaLinhart replied:
You'e right, this is because we might be too discriminative when using ensembling. I will change the number of ensemble models such that the dashed lines are in the gret CR area.

JuliaLinhart replied:
However, I wonder if this is necessary? While I agree that this can be confusing, the theoretical null is not what we compare to in practice.. 

I'm just curious, I agree to keep the larger CR including the true null for the tutorial.

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+    "    axes[i][2].set_title(\"marginal 2 \")\n",
+    "\n",
+    "    for j in range(3):\n",
+    "        axes[j][0].set_ylabel(f\"obsevation {j + 1}\")\n",
```
```suggestion
    "        axes[j][0].set_ylabel(f\"observation {j + 1}\")\n",
```

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+    "        color=\"gray\"\n",
+    "    )\n",
+    "\n",
+    "    axes[i][1].set_title(\"marginal  1\")\n",
+    "    axes[i][2].set_title(\"marginal 2 \")\n",
```
```suggestion
    "    axes[0][1].set_title(\"marginal 1\")\n",
    "    axes[0][2].set_title(\"marginal 2\")\n",
```
These two lines can also move outside of the for loop.

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+   "source": [
+    "## The case of Normalizing Flows ($\\ell$-C2ST-NF)\n",
+    "\n",
+    "$\\ell$-C2ST can also be specialized for normalizing flows,leading to improved test performance. The idea is to train and evaluate the classifiers in the space of the base distribution of the normalizing flow, instead of the parameter space that can be highly structures. \n",
```
```suggestion
    "$\\ell$-C2ST can also be specialized for normalizing flows, leading to improved test performance. The idea is to train and evaluate the classifiers in the space of the base distribution of the normalizing flow, instead of the parameter space that can be highly structured. \n",
```

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+    "## The case of Normalizing Flows ($\\ell$-C2ST-NF)\n",
+    "\n",
+    "$\\ell$-C2ST can also be specialized for normalizing flows,leading to improved test performance. The idea is to train and evaluate the classifiers in the space of the base distribution of the normalizing flow, instead of the parameter space that can be highly structures. \n",
+    "Indeed, following Theorem 4 of [[Linhart et al., 2023]](https://arxiv.org/abs/2306.03580), the null hypothesis $\\mathcal{H}_0(x_\\mathrm{o}) := q_\\phi(\\theta\\mid x_\\mathrm{o}) = p(\\theta \\mid x_\\mathrm{o})$ of *local consistency* holds if, and only if, the inverse flow transformation applied to the target distribution recovers the base distribution. This gives us the following new nu ...
```
```suggestion
    "Following Theorem 4 of [[Linhart et al., 2023]](https://arxiv.org/abs/2306.03580), the null hypothesis $\\mathcal{H}_0(x_\\mathrm{o}) := q_\\phi(\\theta\\mid x_\\mathrm{o}) = p(\\theta \\mid x_\\mathrm{o})$ of *local consistency* holds if, and only if, the inverse flow transformation applied to the target distribution recovers the base distribution. This gives us the following new null hypothesis for posterior estimators based on normalizing flows (cf. Eq. 17 in [[Linhart et al., 2023]](https://arxiv.org/abs/2306.03580)):\n",
```

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+    "\n",
+    "which leads to a new binary classification framework to discriminate between the joint distributions $\\mathcal{N}(0, \\mathbf{I}_m)p(x)$ (class $C=0$) and $p(T_\\phi^{-1}(\\theta ; x_\\mathrm{o}), x_\\mathrm{o})$ (class $C=1$).\n",
+    "\n",
+    "This comes with two main advantages leading to a statistically more powerful and flexible test: \n",
```
```suggestion
    "This results in two main advantages leading to a statistically more performant and flexible test: \n",
```
`power` has a double meaning in hypotheses testing.

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+    "- no permutation method is used inside `train_under_null_hypothesis`\n",
+    "\n",
+    "\n",
+    "> Note: **Quick reminder on Normalizing Flows.** We consider a **conditional Normalizing Flow** $q_{\\phi}(\\theta \\mid x)$ with base distribution $p(z) = \\mathcal{N}(0,\\mathbf{1}_m)$ and bijective transormation $T_{\\phi}(.; x)$ defined on $\\mathbb{R}^2$ and for all $x \\in \\mathbb{R}^2$. Sampling from the normalizing flow consists in applying the forward transformation:\n",
```
```suggestion
    "> Note: **Quick reminder on Normalizing Flows.** We consider a **conditional Normalizing Flow** $q_{\\phi}(\\theta \\mid x)$ with base distribution $p(z) = \\mathcal{N}(0,\\mathbf{1}_m)$ and bijective transormation $T_{\\phi}(.; x)$ defined on $\\mathbb{R}^2$ and for all $x \\in \\mathbb{R}^2$ for our example problem in 2D. Sampling from the normalizing flow consists of applying the forward transformation $T_{\\phi}$:\n",
```

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+    "\n",
+    "> Note: **Quick reminder on Normalizing Flows.** We consider a **conditional Normalizing Flow** $q_{\\phi}(\\theta \\mid x)$ with base distribution $p(z) = \\mathcal{N}(0,\\mathbf{1}_m)$ and bijective transormation $T_{\\phi}(.; x)$ defined on $\\mathbb{R}^2$ and for all $x \\in \\mathbb{R}^2$. Sampling from the normalizing flow consists in applying the forward transformation:\n",
+    ">$$\\theta = T_{\\phi}(z; x) \\sim q_{\\phi}(\\theta \\mid x), \\quad z\\sim p(z)~.$$\n",
+    ">**Characterization of the null hypothesis.** Comparing the estimated and true posterior distributions is equivalent to comapring the base distribtuion to the inverse transformed prior samples: \n",
```
```suggestion
    ">**Characterization of the null hypothesis.** Comparing the estimated and true posterior distributions is equivalent to comparing the base distribution to the inversely transformed prior samples: \n",
```

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+   "source": [
+    "### Set up $\\ell$-C2ST-NF\n",
+    "\n",
+    "The setup of the NF-version is the same as for the original $\\ell$-C2ST, but the trained classifiers can bve used to compute test results and diagnostics for any new observation **and new posterior estimator**."
```
```suggestion
    "The setup of the NF version is the same as for the original $\\ell$-C2ST, but the trained classifiers can be used to compute test results and diagnostics for any new observation **and new posterior estimator**."
```

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+    "#### Quantitative diagnostics\n",
+    "Same as before: we compute test statistics, confidence regions and p-values.\n",
+    "\n",
+    "**Results:** Again the test is rejected for all threee observations."
```
```suggestion
    "**Results:** Again the test hypothesis is rejected for all three observations."
```

psteinb commented on tutorials/18_diagnostics_lc2st.ipynb:
```diff
@@ -0,0 +1,821 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA/wAAAFRCAYAAADJvKbiAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuNSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/xnp5ZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAB0UElEQVR4nO3deVxUVeM/8M8wMMMmILIrIuKCqbiL5K4koGkqphbfklIxw0p9NKOn3BVTS9NQy6cgC9MstbTS0ERbwBQzc0mRcEsWNyBA1jm/P/xxc2SRWWCG4fN+ve7LmXvPPfcchvnIuatMCCFARERERERERCbFzNANICIiIiIiIiL944CfiIiIiIiIyARxwE9ERERERERkgjjgJyI ...
```
change PP-Plots into `P-P plots`, please! 
> The github webpage wouldn't accept me making a direct suggestion.

psteinb replied:
I suggest to change this line into:
`axes[-1].legend(bbox_to_anchor=(1.1, .5), loc='center left')`

psteinb replied:
change this into `axes[-1].legend(bbox_to_anchor=(1.1, .5), loc='center left')` outside of loop.

psteinb reviewed (changes requested):
This is an awesome PR. Thanks @JuliaLinhart  and all reviewers so far. I felt very humble reviewing it as a lot of dedication, discipline and rigor went into it. 

I focused on reviewing the tutorial. Note, I couldn't directly suggest edits to the notebook towards the latter quarter of the notebook, for some reason the github webpage always wanted to remove images whenever I made a code suggestion.

JuliaLinhart commented:
There you go @psteinb :) Thanks a lot for your review!

janfb commented on sbi/diagnostics/lc2st.py:
```diff
@@ -0,0 +1,786 @@
+                trained_clfs.append(clf_n)
+
+        # ensembling
+        if self.num_ensemble > 1:
+            trained_clfs = []
```
if both `self.num_folds` and `self.num_ensemble` are  `>1` then this effectively resets `trained_clfs = []` and discards the trained classifiers from above, no?

I am wondering what is the difference between doing a k-fold CV and ensembling (except that CV takes different train-test splits?)  do we really need the ensembling here?

JuliaLinhart replied:
Oh you might be right! Let me check! 

Ensembling reduces the variance coming from the classifier initialization (with different seeds). I stubled upon an example where this is quite important.

JuliaLinhart replied:
I'm just wondering if we should allow to do CV and ensembling together (i.e. ensembling inside the CV loop) or if this should raise an error.

janfb replied:
maybe we can just add seeding to the CV? This then should be the same as doing ensembling except that CV iterates on different test and train splits?

JuliaLinhart replied:
Yes, but for small datasets this might not be appropriate. I don't know... Maybe @psteinb can give his opinion on that?

JuliaLinhart replied:
cross val makes different data splits so this would need to be integrated in the null hypothesis I think...

janfb reviewed (commented):
thanks for looking into the classifier performance and variance! 

There is one open question from my previous review and one question regarding the ensemble training. 

Thanks! 🙏

psteinb reviewed (approved):
The tutorial looks good to me!

JuliaLinhart commented:
Hello everyone. I propose in this last commit a solution to the ensembling / cross-val issue stated above. 

- I chose to create a `EnsembleClassifier` class whose predicted probabilities are the average prediction over all classifiers (that differ only by their `random_state`). Cross-val is just a training strategy that can also be performed on an ensemble classifier. 

> Here's  something to think about: The cross-val scores are the test statistics obtained for each fold. If someone wishes to perform a test, i.e. compute p-values, with the cross-val strategy, the test statistic becomes the **average statistic** over all folds. This is different from ensembling, where the test statistic is computed on the **average prediction**. 

- I also added a section **## Classifier choice and calibration data size: how to ensure meaningful test results** in the tutorial if you want to check it out @psteinb .

Finally, I added a small description for the `LC2ST_NF` to be more explicit on what the `theta_o` are. @janfb let me know if that answers your questions.

janfb reviewed (approved):
Thanks @JuliaLinhart for these additional changes and for the commitment during this long PR! 👏 
It all looks good to me now! 

Thanks also @psteinb for your review on this. 

Will be merged soon 🎉

JuliaLinhart commented:
Thank you for your valuable comments and reviews !!
